main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -64,11 +69,40 @@ object Lexer {
 
   private def addToken(token: Token)(implicit s: State): Unit = ???
 
-  private class State {
+  // Converts a `Token` of kind `TokenKind.Err` into a CompilationMessage.
+  // NOTE: Why is this necessary?
+  // We would like the lexer to capture as many errors as possible before terminating.
+  // To do this, the lexer will produce error tokens instead of halting,
+  // each holding a kind of the simple type `ErrKind`.
+  // So we need this mapping to produce a `CompilationMessage`, which is a case class, if there were any errors.
+  private def tokenErrToCompilationMessage(t: Token)(implicit s: State): Option[CompilationMessage] = {
```

Comment:
Suggestion: Change signature to `Token.Err` then change return type to CompilationMessage.

Then where you use this, you can write `tokens.collect { case Token.Err => }`, from this you can construct a `Failure` object.

Reply:
I don't think this is possible. `Token` holds a `TokenKind` which can then be an `TokenKind.Err`. But we need the rest of the data on `Token` as well (`text`, `line` and `col`). Does scala have a way to express the type "Any token where its `kind` field is an instance of `TokenKind.Err`" ?

Reply:
Yes, like a deep pattern match?

```
case Token(TokenKind.Err(...), ..., ...) => 
```